Fix nested dict merges where None appear on one side

epcim opened the pull request:
to be evaluated
possibly does similar or nearly the same as #31 

@AndrewPickford pls review, especially to #1 extent. Thanks


The before behavior:

```
#Dict merges with None and { '_class': {'second': Value(ScaItem('second') resulted in ValueLists like :
{ '_class': {'second': ValueList([Value(ScaItem('second')), Value(ScaItem(None))])}}
```

AndrewPickford commented:
@epcim I'm not following the example, it looks like you have a scalar value 'second' and then a second scalar value of None to merge over the first. Could you write it out in yaml so it's clearer.

For #31, the original reclass model works for scalar references and in handling referenced dictionary and list merges correctly the assumption of only having to overwrite in Parameters._merge_dict breaks. For this an overwrite flag needs to be available in ValueList.render.

epcim commented:
[**@AndrewPickford**](https://github.com/AndrewPickford) commented on [Apr 24, 2018, 11:05 AM GMT+2](https://github.com/salt-formulas/reclass/pull/33#issuecomment-383859404 "2018-04-24T09:05:39Z - Replied by Github Reply Comments"):
> @epcim I'm not following the example, it looks like you have a scalar value 'second' and then a second scalar value of None to merge over the first. Could you write it out in yaml so it's clearer.
> 
> For #31, the original reclass model works for scalar references and in handling referenced dictionary and list merges correctly the assumption of only having to overwrite in Parameters._merge_dict breaks. For this an overwrite flag needs to be available in ValueList.render.

@AndrewPickford pls check this branch https://github.com/salt-formulas/reclass/tree/fix-wrapping-with-debug (in last commit I added tar with my /etc/reclass and debug info in code.

Run with:
```
python ./reclass.py --nodeinfo dontpanic
````

It will output:

```
➜  reclass-new git:(fix-wrapping-with-debug) ✗ python ./reclass.py --nodeinfo dontpanic           
DDD merge_base + descent Entity(Classes([]), Applications([], '~'), Parameters({'_class': {'second': Value(ScaItem('second')), 'env': {'override': Value(ScaItem('env.dev'))}, 'first': Value(ScaItem('first'))}}), Exports({}), uri='yaml_fs:///etc/reclass/classes/classes.yml', name='classes', environment=None)
DDD merge_base + descent Entity(Classes([]), Applications([], '~'), Parameters({'_class': {'second': Value(ScaItem('second')), 'env': {'override': Value(ScaItem('env.dev'))}, 'first': Value(ScaItem('first'))}}), Exports({}), uri='yaml_fs:///etc/reclass/classes/classes.yml', name='classes', environment=None)
DDD merge_base + descent Entity(Classes([]), Applications([], '~'), Parameters({'lab1': {'name': Value(ScaItem('lab1'))}, '_class': {'second': Value(ScaItem('second')), 'env': {'override': Value(ScaItem('env.dev'))}, 'first': Value(ScaItem('first'))}, 'lab': {'name': Value(ScaItem('test')), 'label': Value(ScaItem('jedna'))}}), Exports({}), uri='yaml_fs:///etc/reclass/classes/first.yml', name='first', environment=None)
DDD merge_base + descent Entity(Classes(['classes', '${_class:first}']), Applications([], '~'), Parameters({'_param': {'aaa': Value(RefItem([ScaItem('_param:yyy')])), 'ccc': Value(RefItem([ScaItem('_param:aaa')]))}, 'lab': {'name': Value(ScaItem('test')), 'label': ValueList([Value(ScaItem('jedna')), Value(ScaItem('dva'))])}, 'lab2': {'name': Value(ScaItem('lab2'))}, 'lab1': {'name': Value(ScaItem('lab1'))}, '_class': {'second': ValueList([Value(ScaItem('second')), Value(ScaItem('second'))]), 'env': {'override': ValueList([Value(ScaItem('env.dev')), Value(ScaItem('env.dev'))])}, 'first': ValueList([Value(ScaItem('first')), Value(ScaItem('first'))])}, 'mykey': Value(RefItem([ScaItem('_param:ccc')]))}), Exports({}), uri='yaml_fs:///etc/reclass/classes/second.yml', name='second', environment=None)
**[WARNING] Reclass class not found: 'lab.ValueList([Value(ScaItem('env.dev')), Value(ScaItem('env.dev'))])'. Skipped!**
DDD merge_base + descent Entity(Classes(['classes', '${_class:first}', '${_class:second}', 'lab.${_class:env:override}']), Applications([], '~'), Parameters({'mykey2': {'tree': {'to': {'fail': Value(RefItem([ScaItem('_param:kkk')]))}}}, '_param': {'aaa': ValueList([Value(RefItem([ScaItem('_param:yyy')])), Value(ScaItem(999))]), 'ccc': ValueList([Value(RefItem([ScaItem('_param:aaa')])), Value(ScaItem(444))]), 'kkk': Value(ScaItem(555))}, 'mkkek3': {'tree': {'to': {'fail': Value(RefItem([ScaItem('_param:kkk')]))}, 'another': {'xxxx': Value(RefItem([ScaItem('_param:kkk')]))}}}, 'lab': {'name': Value(ScaItem('test')), 'label': ValueList([Value(ScaItem('jedna')), Value(ScaItem('dva')), Value(ScaItem('tri'))])}, 'lab3': {'name': Value(ScaItem('lab3'))}, 'lab2': {'name': Value(ScaItem('lab2'))}, 'lab1': {'name': Value(ScaItem('lab1'))}, '_reclass_': {'environment': Value(ScaItem('base')), 'name': {'full': Value(ScaItem('dontpanic')), 'short': Value(ScaItem('dontpanic'))}}, '_class': {'second': ValueList([Value(ScaItem('second')), Value(ScaItem('second'))]), **'env': {'override': ValueList([Value(ScaItem('env.dev')), Value(ScaItem('env.dev'))])}**, 'first': ValueList([Value(ScaItem('first')), Value(ScaItem('first'))])}, 'mykey': ValueList([Value(RefItem([ScaItem('_param:ccc')])), Value(RefItem([ScaItem('_param:ccc')]))])}), Exports({}), uri='yaml_fs:///etc/reclass/classes/third.yml', name='third', environment=None)
[WARNING] Reference '${_param:yyy}' undefined
__reclass__:
  environment: base
  name: dontpanic
  node: ./dontpanic
  timestamp: Wed Apr 25 08:42:03 2018
  uri: yaml_fs:///etc/reclass/nodes/dontpanic.yml
applications: []
classes:
- classes
- ${_class:first}
- ${_class:second}
- lab.${_class:env:override}
- third
environment: base
exports: {}
parameters:
  _class:
    env:
      override: env.dev
    first: first
    second: second
  _param:
    aaa: 999
    ccc: 444
    kkk: 555
  _reclass_:
    environment: base
    name:
      full: dontpanic
      short: dontpanic
  lab:
    label: tri
    name: test
  lab1:
    name: lab1
  lab2:
    name: lab2
  lab3:
    name: lab3
  mkkek3:
    tree:
      another:
        xxxx: 555
      to:
        fail: 555
  mykey: 444
  mykey2:
    tree:
      to:
        fail: 555
```

Mind, where I inserded "**" to highlifht.

A,
```
[WARNING] Reclass class not found: 'lab.ValueList([Value(ScaItem('env.dev')), Value(ScaItem('env.dev'))])'. Skipped!
```

B, 
```
**'env': {'override': ValueList([Value(ScaItem('env.dev')), Value(ScaItem('env.dev'))])}**
```

and multiple occurrences in debug - where ValueList(a,a) - ie: contains the same. Similarly it did a ValueList of "a" and "None" (which was supposed  to be only Vaue(a) or Value(None)).

The branch has reverted fix. Don't pay attention to "global-classes" commit in history (it's there only to trigger the behaviour). 

I understand that `if klass not seen` is modified and we might get loop not originally expected as I parse+merge the same class again. 

The above is about a part when _dict_merge on it's leafs have the same objects, but returns them as ValueList instead of one Value.

fixed with right part at:
```
if not value_list and (new.contents() is None or new.contents() == cur.contents()):
```

Now where I got None. Well I originally tried to pre-define merge_base with some "global" parameters, before I came with the idea of global-classes. One solution was to keep `_class` key None thus will be always overriden:

```
--- a/reclass/core.py
+++ b/reclass/core.py
@@ -111,7 +111,8 @@ class Core(object):
             environment = self._settings.default_environment
 
         if merge_base is None:
-            merge_base = Entity(self._settings, name='empty (@{0})'.format(nodename))
+            p = Parameters({'_class': {'first':None}}, self._settings, '')
+            merge_base = Entity(self._settings, parameters=p, name='empty (@{0})'.format(nodename))
```

On my fix, the change:
```
-        if cur is None:
+        if cur is None or str(cur) == 'None':
```
was required because for some reason cur was Value(ScaItem('None')) and thus reason for the right part.

Now I consider to add "global-class" feature as we have minor, but still an use case. Anyway the "fix" smells to me more like an workaround (at least the part `str(cur) == 'None'`). I would appreciate if you would spend a minute to put some light on it.

AndrewPickford commented:
I think I understand what's going on here, for:

> [WARNING] Reclass class not found: 'lab.ValueList([Value(ScaItem('env.dev')), Value(ScaItem('env.dev'))])'. Skipped!

there are two components: the formation of the ValueList and then the rendering of the ValueList as part of setting the klass variable.

The ValueList is formed due to repeatedly processing the global classes with:

`if klass not in seen or self._gcl_r.match(klass):`

in _recurse_entity. The individual keys are then merged on top of each other to form the ValueLists with repeated elements.

When the klass name is subsequently rendered with:

`klass = str(self._parser.parse(klass, self._settings).render(merge_base.parameters.as_dict(), {}))`

the render method returns a string representation of it's contents. For references this is the string representation of the referenced value which for the ValueList here is 'ValueList([Value(ScaItem('env.dev'), Value(ScaItem('env.dev'))])'. The merging of the ValueList entries on top of each other never happens.

The rendering for scalar values like Value(ScaItem('second')) works as here the render function returns the string 'second'.

In order to correctly render a reference to a key which is a ValueList the ValueList would first need to be rendered. The render method does not call the render method on the keys on which the key being rendered relies on. Instead correct rendering order is handled by the Parameters.interpolate method. 

Essentially it's the global class implementation that's causing the problem. For global classes is the idea to load the same class file or files for all nodes? There should be another way to do that.

For the `if cur is None or str(cur) == 'None':` change, in normal processing the the cur variable will be a Value object. All the key values in the dictionary passed into the Parameters.merge method are in normal processing during the _recurse_entity processing wrapped with value objects. The `str(cur) == 'None'` works as keys with a None value are wrapped into Value objects as Value(Scalar(None)). The string representation of such a Value object is 'None'.

AndrewPickford commented:
I'm not following your question. If there is a ValueList(Value(xyz), Value(None)) currently if xyz is a scalar then it will get overwritten by None otherwise it's an error. I think we are talking at cross purposes here.

For adding parameters before the merge starts they should go into the Core._node_entity method. This how the \_reclass\_ parameters get added with the line:

`base_entity.merge_parameters(self._get_automatic_parameters(nodename, node_entity.environment))`

A value of None (not the string but the python None) can indeed come from yaml:
```yaml
parameters:
  test:
```
would set test to None.

I'm confident that the rendering of ValueLists correctly is only an issue in Core._recurse_entity. Barring some bug they are rendered correctly during the interpolation step. It's a matter of getting the rendering order correct. That's done in Parameters._interpolate_references were all keys that a given key depends on (ie the keys that the key being evaluated references) are rendered first. This ensures that if a reference points to a key that is a ValueList then the ValueList will be rendered first and the key being referenced will then contain a int, string, dict etc which is used to set the value of the referencing key.

There's one bug that I know of here. Currently if you have one escaped string overwrite a second escaped string, ie:
```yaml
#class1.yml
parameters:
  a: \${not_a_ref}

#class2.yml
parameters:
  a: \${also_not_a_ref}
```
this fails. The ValueList created can not be evaluated. I'll put in a pull request with a fix for this, hopefully tomorrow, which also tidies up the Parameter.merge method and simplifies how dictionary merges handle escaped strings.

epcim commented:
OK. I got the answer in this post. Closing.
(Again, thank you for an explanation, will help in future debuging;)
